rclcpp/test/executors/test_multi_threaded_executor.cpp

```diff
@@ -89,10 +89,9 @@ TEST_F(TestMultiThreadedExecutor, timer_over_take) {
         double diff = std::abs((now - last).nanoseconds()) / 1.0e9;
         last = now;
 
-        if (diff < PERIOD - TOLERANCE || diff > PERIOD + TOLERANCE) {
+        if (diff < PERIOD - TOLERANCE) {
           executor.cancel();
           ASSERT_GT(diff, PERIOD - TOLERANCE);
```

Comment:
this is first checking `diff < UPPER_BOUND`, and then asserting `diff > UPPER_BOUND`? (with `UPPER_BOUND = PERIOD - TOLERANCE`)
how the second condition can happen after the first check?

this is failing sometimes in CI: https://ci.ros2.org/view/nightly/job/nightly_linux_repeated/1858/testReport/junit/rclcpp/TestMultiThreadedExecutor/timer_over_take/.
I'm not sure why, maybe a floating point rounding error?

Reply:
If I understand my previous comment (https://github.com/ros2/rclcpp/pull/907#discussion_r339241075), I think the intention is that if `diff < UPPER_BOUND`, then we want the test fail. I admit it is confusing, and we could probably clarify with a comment.

Reply:
oh, so in that case we want the test to unconditionally fail, right?
I think that either adding a comment or using `FAIL()` would be more clear (now that we're checking only one of the conditions).

Reply:
Makes sense to me.

Regarding the test failure, I'm still not sure if the bug relies in the test code or not. Having timing dependent tests are tricky.

Reply:
> Regarding the test failure, I'm still not sure if the bug relies in the test code or not. Having timing dependent tests are tricky.

I have taken a look, and it seems to be a flaky test (though the timer is triggered MUCH BEFORE than expected).
I've read the original PR https://github.com/ros2/rclcpp/pull/383 that motivated these changes, the only change I see in the multithreaded executor since then is https://github.com/ros2/rclcpp/pull/836, which doesn't seem to be related.

The test seems to be super flaky. I have checked the last few repeated jobs, and it has only failed once.
IMO, it sounds like we can ignore it.

@wjwwood do you think that any of the reset changes can be related to this? or is it just a flaky test?

Reply:
> @wjwwood do you think that any of the reset changes can be related to this? or is it just a flaky test?

No, not that I can think of. Nothing about the executor really changed. I didn't get to that part yet.

Reply:
>The test seems to be super flaky. I have checked the last few repeated jobs, and it has only failed once.
IMO, it sounds like we can ignore it.

Maybe that's the case for `master`, but on `eloquent` it fails more than it succeeds:

http://build.ros2.org/job/Eci__nightly-connext_ubuntu_bionic_amd64/307/testReport/rclcpp/TestMultiThreadedExecutor/timer_over_take/history/

I'd like to either remove this line or add an `xfail` label to it. What would you prefer?

Reply:
I think we should improve the test, instead of that, to check if there's actually a regression or not.
Incrementing both period and tolerance by an order of magnitude should make the test reliable (if there's not a real regression).

I will create a PR today/tomorrow-ish.

Reply:
I proposed a PR increasing period/tolerance by an order of magnitude, and it failed too https://github.com/ros2/rclcpp/pull/1105#issuecomment-624757046.

There seems to be a real bug, so it should be triagged/fixed.

Reply:
It might be connected with this other failure too https://github.com/ros2/rcl/pull/640.

Reply:
> I proposed a PR

Thanks!

> it should be triagged/fixed.

Until it's fixed, the test should be either commented out or marked as `xfail` so it stops showing on CI results. What do you think?

Reply:
I don't know if `xfail` is intended to be used only on flaky tests, or also in tests that are failing due to something being actually broken.

If that mark is also intended for the second case, then it's ok to do it.

Reply:
> If that mark is also intended for the second case, then it's ok to do it.

I believe so: https://github.com/ros2/ros2/issues/900